Implement Claim a Facility Dashboard

kellyi opened the pull request:
## Overview

This PR implements some MVP facility claim review dashboard pages which admin users can access to review and annotate FacilityClaims and also update statuses. There's a list page at

http://localhost:6543/dashboard/claims

and a detail page for each claim ID at

http://localhost:6543/dashboard/claims/claimID

On the details page, admin users can see data as depicted below and also update statuses and make notes.

On the API side, I added two migrations to store data about reviewing facility claims, added endpoints for approving, denying, and rejecting facility claims as well as adding new notes, set up the needed serializers, and added functionality to email facility claimants on status changes. (The emails for the latter are templates now but we can flesh them out in specific ways later).

I overloaded the facility claim notes field a little bit by using it to log details about status changes too. I did this specifically because I wanted the notes trail to be a log -- and not to have a separate status log.

Connects #517 

## Demo

![Screen Shot 2019-06-06 at 5 54 20 PM](https://user-images.githubusercontent.com/4165523/59069242-8c27f400-8884-11e9-9fc0-47058d84c20c.png)

![Screen Shot 2019-06-06 at 5 55 28 PM](https://user-images.githubusercontent.com/4165523/59069314-be395600-8884-11e9-9224-752fe2e4fa1c.png)

![Screen Shot 2019-06-06 at 5 56 12 PM](https://user-images.githubusercontent.com/4165523/59069323-c1ccdd00-8884-11e9-8da1-be91c71279e4.png)

![Screen Shot 2019-06-06 at 5 56 19 PM](https://user-images.githubusercontent.com/4165523/59069331-c72a2780-8884-11e9-8067-413d6e1cf60f.png)

![Screen Shot 2019-06-06 at 5 56 35 PM](https://user-images.githubusercontent.com/4165523/59069337-cb564500-8884-11e9-9d37-98138ee04cfe.png)

## Testing Instructions

- get this branch then `resetdb`, `processfixtures` and `createsuperuser`
- `./scripts/server`
- `./scripts/manage createsuperuser` then sign in to the Django Admin as your super user an switch on `claim_a_facility`
- register for a new account in the client, then verify it
- submit a couple of claim a facility forms, making the test submitted in each distinct enough to tell them apart
- sign out, then sign back in as your super user and visit http://localhost:6543/dashboard/claims and verify you see the facility claims table
    - clicking a one of the obvious links here should route you either to user profile or facility details pages
    - clicking a row *not* on a link should take you to the details page
- refresh a details page and verify that it loads the details
- try out adding some notes and verify that it works as expected
- try out approving and denying and verify that that works too
- try out revoking an approved facility and verify that that also works

## Checklist

- [x] `fixup!` commits have been squashed
- [x] CI passes after rebase
- [x] CHANGELOG.md updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines

kellyi commented:
So far:

![Screen Shot 2019-06-04 at 4 51 18 PM](https://user-images.githubusercontent.com/4165523/58913132-89dd6280-86e9-11e9-83ab-a6868d6d9910.png)

kellyi commented on src/django/api/models.py:
```diff
@@ -422,6 +422,66 @@ class FacilityClaim(models.Model):
+    note = models.TextField(
+        null=False,
+        blank=False,
+        help_text='The review note')
```
This needs a `history = HistoricalRecords()` setting, too.

kellyi replied:
I checked and it's there -- https://github.com/open-apparel-registry/open-apparel-registry/blob/1116b620d3870d8b92d6b4da81588bf91dc85705/src/django/api/models.py#L489 -- but just not showing up in the diff.

kellyi commented:
It occurred to me that this and the other claim a facility PR need to have the Django admin updated to include the history list display, so I'll push another commit here to do that.

jwalgran commented:
Looking at this now

jwalgran commented:
The merge conflict is just simple overlaps.

jwalgran commented:
There is a migration conflict with develop. I think it can be resolved by rebasing and adding         
```
('api', '0009_contributor_add_verified_fields_and_history_20190605'),
```
as an additional dependency to the 0010 migration.

```
CommandError: Conflicting migrations detected; multiple leaf nodes in the migration graph: (0009_contributor_add_verified_fields_and_history_20190605, 0010_facilityclaimreviewnote_historicalfacilityclaimreviewnote in api).
To fix them run 'python manage.py makemigrations --merge'
```

jwalgran commented on src/app/src/components/DashboardClaimDetailsControls.jsx:
```diff
@@ -0,0 +1,328 @@
+                                        color="secondary"
+                                        onClick={dialogContentData.action}
+                                    >
+                                        Yes, {dialogContentData.actionTerm}
```
Consider dropping the "Yes," prefix so that the button just has the action, which I find to be more clear.

kellyi replied:
Good point -- this was leftover from a prior version where the "Cancel" button was titled stuff like "No, do not approve".

kellyi commented on src/django/api/models.py:
```diff
@@ -422,6 +422,66 @@ class FacilityClaim(models.Model):
         choices=STATUS_CHOICES,
         default=PENDING,
         help_text='The current status of this facility claim')
+    approval_reason = models.TextField(
```
That makes sense to me. I split these into separate fields before I decided to also log the status changes to the facility claim notes log, so the idea was to make it easy to see at a glance what all the status changes that had been made to a claim. Now that they're logged, though, I think it's okay to fallback to looking these up in the historical values when necessary.

kellyi commented on src/django/api/templates/mail/claim_facility_approval_body.html:
```diff
@@ -0,0 +1,35 @@
+            </li>
+        </ul>
+        <p>
+            Here's the reason for the approval:
```
Sounds good. I applied this change also to the denial and revocation reason templates, such that the reason section won't render now if the reason is an empty string.

kellyi commented on src/django/api/views.py:
```diff
@@ -1289,13 +1295,180 @@ class FacilityClaimViewSet(viewsets.ModelViewSet):
+            claim = FacilityClaim.objects.get(pk=pk)
+
+            if claim.status != FacilityClaim.PENDING:
+                raise APIException('Bad request')
```
:+1:

kellyi commented on src/django/api/views.py:
```diff
@@ -1289,13 +1295,180 @@ class FacilityClaimViewSet(viewsets.ModelViewSet):
+            claim = FacilityClaim.objects.get(pk=pk)
+
+            if claim.status != FacilityClaim.PENDING:
+                raise APIException('Bad request')
```
:+1:

kellyi commented on src/django/api/views.py:
```diff
@@ -1289,13 +1295,180 @@ class FacilityClaimViewSet(viewsets.ModelViewSet):
+            claim = FacilityClaim.objects.get(pk=pk)
+
+            if claim.status != FacilityClaim.APPROVED:
+                raise APIException('Bad request')
```
:+1:

kellyi commented:
I rebased this on develop and squashed in some changes to the existing commit:

- the dashboard `AppGrid` title now renders different text depending on the page, thanks to having changed it to accept a `node`
- the facility and contributor fields are now links
- I added `white-space: pre-line` to ensure text entered via textareas now is displayed with proper linebreaks
- the FacilityClaim model now has fields for `status_change...` which are used to record the most recent status change
- the email templates now render the reason conditionally

This is ready for another look @jwalgran

jwalgran commented:
I just noticed there are several `return NotFound()` and `return PermissionDenied()`. It appears that we need to raise these Exception subclasses rather than return them.

<img width="687" alt="Screen Shot 2019-06-10 at 10 15 27 AM" src="https://user-images.githubusercontent.com/17363/59213002-b3264480-8b68-11e9-8c95-bdeced13345c.png">

kellyi commented:
Good catch -- fixed these in 7345749c1480f18d37768de9adc8409ea6e76e9f

jwalgran commented:
Looking at this now

kellyi commented:
Rebased on develop & squashed.

jwalgran commented on src/django/api/templates/mail/claim_facility_denial_body.html:
```diff
@@ -0,0 +1,37 @@
+                Facility URL: {{ facility_url }}
+            </li>
+        </ul>
+        {% if len(denial_reason) %}
```
I am running into a `Could not parse the remainder: '(denial_reason)' from 'len(denial_reason)'` exception when the deny logic. I don't think we have access to `len` in a template. We likely have to use the `|length` filter instead

```diff
diff --git a/src/django/api/templates/mail/claim_facility_denial_body.html b/src/django/api/templates/mail/claim_facility_denial_body.html
index 6bef756..04da155 100644
--- a/src/django/api/templates/mail/claim_facility_denial_body.html
+++ b/src/django/api/templates/mail/claim_facility_denial_body.html
@@ -21,7 +21,7 @@
                 Facility URL: {{ facility_url }}
             </li>
         </ul>
-        {% if len(denial_reason) %}
+        {% if denial_reason|length %}
         <p>
             Here's the reason for the denial:
         </p>
diff --git a/src/django/api/templates/mail/claim_facility_denial_body.txt b/src/django/api/templates/mail/claim_facility_denial_body.txt
index ea32f30..ad13953 100644
--- a/src/django/api/templates/mail/claim_facility_denial_body.txt
+++ b/src/django/api/templates/mail/claim_facility_denial_body.txt
@@ -6,7 +6,7 @@ You're receiving this email because we denied your request to claim this Open Ap
 - Facility: {{ facility_name }}, {{ facility_address }}
 - Facility URL: {{ facility_url }}
 
-{% if len(denial_reason) %}
+{% if denial_reason|length %}
 Here's the reason for the denial:
 
 {{ denial_reason }}

```

```
ERROR:django.request:Internal Server Error: /api/facility-claims/1/deny/
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/django/template/smartif.py", line 175, in translate_token
    op = OPERATORS[token]
KeyError: 'len(denial_reason)'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py", line 128, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py", line 126, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.7/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/rest_framework/viewsets.py", line 116, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/rest_framework/views.py", line 495, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.7/site-packages/rest_framework/views.py", line 455, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.7/site-packages/rest_framework/views.py", line 492, in dispatch
    response = handler(request, *args, **kwargs)
  File "/usr/local/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "/usr/local/src/api/views.py", line 1409, in deny_claim
    send_claim_facility_denial_email(request, claim)
  File "/usr/local/src/api/mail.py", line 72, in send_claim_facility_denial_email
    text_template = get_template('mail/claim_facility_denial_body.txt')
  File "/usr/local/lib/python3.7/site-packages/django/template/loader.py", line 15, in get_template
    return engine.get_template(template_name)
  File "/usr/local/lib/python3.7/site-packages/django/template/backends/django.py", line 34, in get_template
    return Template(self.engine.get_template(template_name), self)
  File "/usr/local/lib/python3.7/site-packages/django/template/engine.py", line 144, in get_template
    template, origin = self.find_template(template_name)
  File "/usr/local/lib/python3.7/site-packages/django/template/engine.py", line 126, in find_template
    template = loader.get_template(name, skip=skip)
  File "/usr/local/lib/python3.7/site-packages/django/template/loaders/base.py", line 30, in get_template
    contents, origin, origin.template_name, self.engine,
  File "/usr/local/lib/python3.7/site-packages/django/template/base.py", line 160, in __init__
    self.nodelist = self.compile_nodelist()
  File "/usr/local/lib/python3.7/site-packages/django/template/base.py", line 198, in compile_nodelist
    return parser.parse()
  File "/usr/local/lib/python3.7/site-packages/django/template/base.py", line 483, in parse
    raise self.error(token, e)
  File "/usr/local/lib/python3.7/site-packages/django/template/base.py", line 481, in parse
    compiled_result = compile_func(self, token)
  File "/usr/local/lib/python3.7/site-packages/django/template/loader_tags.py", line 233, in do_block
    nodelist = parser.parse(('endblock',))
  File "/usr/local/lib/python3.7/site-packages/django/template/base.py", line 483, in parse
    raise self.error(token, e)
  File "/usr/local/lib/python3.7/site-packages/django/template/base.py", line 481, in parse
    compiled_result = compile_func(self, token)
  File "/usr/local/lib/python3.7/site-packages/django/template/defaulttags.py", line 955, in do_if
    condition = TemplateIfParser(parser, bits).parse()
  File "/usr/local/lib/python3.7/site-packages/django/template/defaulttags.py", line 888, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/django/template/smartif.py", line 166, in __init__
    mapped_tokens.append(self.translate_token(token))
  File "/usr/local/lib/python3.7/site-packages/django/template/smartif.py", line 177, in translate_token
    return self.create_var(token)
  File "/usr/local/lib/python3.7/site-packages/django/template/defaulttags.py", line 891, in create_var
    return TemplateLiteral(self.template_parser.compile_filter(value), value)
  File "/usr/local/lib/python3.7/site-packages/django/template/base.py", line 568, in compile_filter
    return FilterExpression(token, self)
  File "/usr/local/lib/python3.7/site-packages/django/template/base.py", line 668, in __init__
    "from '%s'" % (token[upto:], token))
django.template.exceptions.TemplateSyntaxError: Could not parse the remainder: '(denial_reason)' from 'len(denial_reason)'
```

kellyi replied:
Fixed in b289c2626f51de30ae1a4fb1d766c94e4434e310

jwalgran commented on src/django/api/serializers.py:
```diff
@@ -373,6 +375,81 @@ def get_country_name(self, facility):
+            }
+
+        return {
+            'status_change_by': claim.status_change_by.email,
```
I hit an exception on this line.

```
ERROR:django.request:Internal Server Error: /api/facility-claims/1/deny/
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py", line 128, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py", line 126, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.7/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/rest_framework/viewsets.py", line 116, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/rest_framework/views.py", line 495, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.7/site-packages/rest_framework/views.py", line 455, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.7/site-packages/rest_framework/views.py", line 492, in dispatch
    response = handler(request, *args, **kwargs)
  File "/usr/local/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "/usr/local/src/api/views.py", line 1411, in deny_claim
    response_data = FacilityClaimDetailsSerializer(claim).data
  File "/usr/local/lib/python3.7/site-packages/rest_framework/serializers.py", line 563, in data
    ret = super(Serializer, self).data
  File "/usr/local/lib/python3.7/site-packages/rest_framework/serializers.py", line 262, in data
    self._data = self.to_representation(self.instance)
  File "/usr/local/lib/python3.7/site-packages/rest_framework/serializers.py", line 530, in to_representation
    ret[field.field_name] = field.to_representation(attribute)
  File "/usr/local/lib/python3.7/site-packages/rest_framework/fields.py", line 1889, in to_representation
    return method(value)
  File "/usr/local/src/api/serializers.py", line 431, in get_status_change
    'status_change_by': claim.status_change_by.email,
AttributeError: 'NoneType' object has no attribute 'email'
```

It looks like it is caused by the fact that the `deny_claim` view is not setting the `status_change_*` fields.

kellyi replied:
I fixed this in 204d9548f31cde4d35f74618dc355f6746b7288a

jwalgran commented on src/django/api/migrations/0011_add_statuschange_fields_to_facilityclaim.py:
```diff
@@ -0,0 +1,155 @@
+# Generated by Django 2.0.13 on 2019-06-10 15:22
```
Can this be combined with the 0010 migration without too much hassle. If it is not too much trouble I think it would be better to not have 2 migrations in the same PR that add a field and then immediately remove it.

kellyi replied:
Yeah, these can be combined.

jwalgran commented on src/django/api/views.py:
```diff
@@ -1280,3 +1288,193 @@ def api_feature_flags(request):
+            if claim.status != FacilityClaim.APPROVED:
+                raise APIException('Only APPROVED claims can be revoked.')
+
+            claim.revocation_reason = request.data.get('reason', '')
```
I believe these need to be replaced by `status_change_*` properties.

kellyi replied:
Fixed in 204d9548f31cde4d35f74618dc355f6746b7288a

jwalgran commented on src/django/api/views.py:
```diff
@@ -1280,3 +1288,193 @@ def api_feature_flags(request):
+            if claim.status != FacilityClaim.PENDING:
+                raise APIException('Only PENDING claims can be denied.')
+
+            claim.denial_reason = request.data.get('reason', '')
```
I believe these need to be replaced by `status_change_*` properties.

kellyi replied:
Fixed up in 204d9548f31cde4d35f74618dc355f6746b7288a

jwalgran commented on src/django/api/views.py:
```diff
@@ -1280,3 +1288,193 @@ def api_feature_flags(request):
+            if claim.status != FacilityClaim.PENDING:
+                raise APIException('Only PENDING claims can be approved.')
+
+            claim.approval_reason = request.data.get('reason', '')
```
I believe these need to be replaced by `status_change_*` properties.

kellyi replied:
Oof -- I fixed this in 204d9548f31cde4d35f74618dc355f6746b7288a

kellyi commented:
I fixed the bugs described above in fixup commits and I also added some tests to prove the endpoints work as they should:

2fb3c858068450d866bfb9217221e2cc8c67ae18

kellyi commented:
OK! Ready for what I hope is a final review @jwalgran

kellyi commented:
Awesome! Thanks for all your help with this!
